### DISCUSSION TITLE
Usage of print ast ql file

### DISCUSSION THREAD

**Discussion by Naman-ntc** [github/securitylab#478] [Q&A] [answered]
Print AST command is used in vscode-codeql extension where it uses the printast query in qlpack and displays the results. 
I wanted to know how to get the ast outside of the vscode-extension?

[printAst.ql](https://github.com/github/codeql/blob/main/javascript/ql/src/printAst.ql) implements a configuration to print ast for a single file (with `selectedSourceFile`) but it is unclear how to use this configuration

**Comment by hvitved** [comment, marked as answer]
This query is specifically written to be executed in the context of VS Code (because of the `external string selectedSourceFile` predicate). If you want to run it externally, you can make your own query:

```ql
import javascript
import semmle.javascript.PrintAst
import definitions
```

will print out the AST for _all_ files, but you could also restrict it

```ql
import javascript
import semmle.javascript.PrintAst
import definitions

private File relevantSourceFile() { ... }

class PrintAstConfigurationOverride extends PrintAstConfiguration {
  override predicate shouldPrint(Locatable e, Location l) {
    super.shouldPrint(e, l) and
    l.getFile() = relevantSourceFile()
  }
}
```

**Reply by Naman-ntc** [reply]
Thanks alot for the prompt responses! I wasn't aware that queries can be written without a select statement or about nodes/edges in bqrs result sets. 
I was able to complete `relevantSourceFile` for a.js in root directory where database is created as follows
```
private File relevantSourceFile() {
  exists(Folder dir | result = dir.getJavaScriptFile("a"))
}
```
In the case of a more complicated directory structure, I suppose I would also need to use `getRelativePath` function to specify the directory? If possible it would be very helpful if you could point to some existing code that might be fetching file from a hardcoded pathstring

**Reply by hvitved** [reply]
You can use `getRelativePath` or `getAbsolutePath` in combination with `matches`, see e.g. https://github.com/github/codeql/blob/main/csharp/ql/test/library-tests/controlflow/graph/Common.qll#L4.

**Reply by Naman-ntc** [reply]
On running the print-ast query and decoding the bqrs with cli - I am able to get a good ast in JSON.  However, the nodes in this json only contain 
* AST node type
* key (semmle.order or semmle.label)
* value (code string)

i.e. it does not provide the location of AST nodes in the javascript file. While I can write some program that matches the nodes from AST JSON file to code in the javascript, I believe some command/function should exist that gives line and column numbers for the nodes as the vs-code extension also links the nodes to the code files.

**EDIT: Also it seems `bqrs interpret` with dot (graph-like output formats) is more informative than `bqrs decode` to a json. Still don't get line and column numbers though**

**Reply by hvitved** [reply]
The following should work:

```
codeql query run -d <db> <query> -o output.bqrs
codeql bqrs decode --entities url --entities string  output.bqrs -o output.txt
```

**Reply by Naman-ntc** [reply]
Oh wow, that works like a charm. Thanks a lot for all the help! 
Why isn't `entities=all` default in `bqrs decode`? This is somewhat confusing behavior

**Reply by hvitved** [reply]
Not sure why the default is what it is, but note that `--entities all` will include an additional `id` column, which is an internal (integer) representation of the entities, so this is mostly useful for internal testing.

**Comment by nakhlarafi** [comment]
Hello there, 

Can I get the AST from the AST viewer in the VScode extension as a file?
